File: server/schemas/tool.py
"""Schemas for Tool Gateway management."""

from __future__ import annotations

from typing import Any
from datetime import datetime

from pydantic import BaseModel


class ToolCreate(BaseModel):
    name: str
    description: str = ""
    category: str = "api"
    endpoint: str = ""
    method: str = "POST"
    input_schema: dict[str, Any] | None = None
    output_schema: dict[str, Any] | None = None
    auth_config: dict[str, Any] | None = None
    timeout_ms: int = 30000
    max_retries: int = 2
    retry_backoff_ms: int = 1000
    is_async: bool = False
    callback_url: str | None = None
    required_permission: str | None = None
    risk_level: str = "info"
    tenant_id: str = "default"


class ToolUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    endpoint: str | None = None
    method: str | None = None
    input_schema: dict[str, Any] | None = None
    output_schema: dict[str, Any] | None = None
    auth_config: dict[str, Any] | None = None
    timeout_ms: int | None = None
    max_retries: int | None = None
    is_async: bool | None = None
    enabled: bool | None = None


class ToolOut(BaseModel):
    id: str
    name: str
    description: str
    category: str
    endpoint: str | None
    method: str
    input_schema: Any
    output_schema: Any
    auth_config: Any
    timeout_ms: int
    max_retries: int
    retry_backoff_ms: int
    is_async: bool
    callback_url: str | None
    required_permission: str | None
    risk_level: str
    tenant_id: str
    enabled: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ToolTestRequest(BaseModel):
    """Test a tool's connectivity."""
    tool_id: str
    test_input: dict[str, Any] | None = None


class ToolTestResponse(BaseModel):
    tool_id: str
    success: bool
    response: Any = None
    error: str | None = None
    latency_ms: float = 0.0

File: console/src/types/tool.ts
export interface Tool {
  id: string;
  name: string;
  description: string;
  category: 'api' | 'function' | 'webhook' | 'rpc';
  endpoint: string;
  method: string;
  input_schema: Record<string, unknown> | null;
  output_schema: Record<string, unknown> | null;
  auth_config: Record<string, unknown> | null;
  timeout_ms: number;
  max_retries: number;
  retry_backoff_ms: number;
  is_async: boolean;
  risk_level: string;
  tenant_id: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface ToolCreate {
  name: string;
  description?: string;
  category?: string;
  endpoint?: string;
  method?: string;
  input_schema?: Record<string, unknown>;
  output_schema?: Record<string, unknown>;
  auth_config?: Record<string, unknown>;
  timeout_ms?: number;
  max_retries?: number;
  risk_level?: string;
  tenant_id?: string;
}

export type ToolUpdate = Partial<ToolCreate>;
